documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/http/server-http-handler-implement.adoc

```diff
@@ -156,6 +156,60 @@ It is therefore mandatory to use `CompletableFuture` APIs that are invoked even
 Failing to do so may result in the `Handler` `callback` parameter to never be completed, causing the request processing to hang forever.
 ====
 
+[[pg-server-http-handler-impl-request-listeners]]
+====== `Request` Listeners
+
+Application may add listeners to the `Request` object to be notified of particular events happening during the request/response processing.
+
+`Request.addIdleTimeoutListener(Predicate<TimeoutException>)` allows you to add an idle timeout listener, which is invoked when an idle timeout period elapses during the request/response processing, if the idle timeout event is not notified otherwise.
+
+The idle timeout listener is invoked only when:
+
+* There is no pending demand for request content; that is, there is not an outstanding call to `Request.demand(Runnable)`.
+* There is no pending response write; that is, there is not an outstanding call to `Response.write(boolean, ByteBuffer, Callback)`.
+
+In the above cases, the application is notified of the idle timeout so there is no need to call the idle timeout listeners.
+In the first case, the demand `Runnable` is invoked and the application will see the idle timeout failure by reading from the `Request`, obtaining a xref:pg-arch-io-content-source[transient failure chunk].
+In the second case, the response write `Callback` will be failed.
```

Comment:
Can we flip this section around so that we describe how the idle timeout is delivered with the bullet point saying what it is.
Something like:

> when an idle timeout occurs it is delivered to the application as follows:
> + If there is pending demand...
> + if there is a pending response...
> + if neither of the above...